src/Arcus.Messaging.Pumps.Abstractions/MessageHandling/MessageHandler.cs

```diff
@@ -0,0 +1,203 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Arcus.Messaging.Abstractions;
+using GuardNet;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Arcus.Messaging.Pumps.Abstractions.MessageHandling
+{
+    /// <summary>
+    /// Represents an abstracted form of the <see cref="IMessageHandler{TMessage,TMessageContext}"/> implementation to handle with different type of generic message and message context types.
+    /// </summary>
+    internal class MessageHandler
+    {
+        private readonly Type _serviceType;
+        private readonly Func<object> _createMessageHandlerImplementation;
+
+        private MessageHandler(Type serviceType, Func<object> createMessageHandlerImplementation)
+        {
+            Guard.NotNull(serviceType, nameof(serviceType));
+            Guard.NotNull(createMessageHandlerImplementation, nameof(createMessageHandlerImplementation));
+
+            _serviceType = serviceType;
+            _createMessageHandlerImplementation = createMessageHandlerImplementation;
+        }
+
+        /// <summary>
+        /// Subtract all the <see cref="IMessageHandler{TMessage,TMessageContext}"/> implementations from the given <paramref name="serviceProvider"/>.
+        /// </summary>
+        /// <param name="serviceProvider">The provided registered services collection.</param>
+        internal static IEnumerable<MessageHandler> SubtractFrom(IServiceProvider serviceProvider)
+        {
+            Guard.NotNull(serviceProvider, nameof(serviceProvider));
+
+            object engine = serviceProvider.GetPropertyValue("Engine");
+            object callSiteFactory = engine.GetPropertyValue("CallSiteFactory", BindingFlags.NonPublic | BindingFlags.Instance);
+
+            var descriptorLookup = (System.Collections.IEnumerable) callSiteFactory.GetFieldValue("_descriptorLookup");
+
+            var messageHandlers = new Collection<MessageHandler>();
+            foreach (object lookup in descriptorLookup)
+            {
+                var serviceType = (Type) lookup.GetPropertyValue("Key");
+                
+                // TODO: update to more strict message handler filter.
```

Comment:
Let's add an issue and remove the TODO

Reply:
Done: https://github.com/arcus-azure/arcus.messaging/issues/68